luigi/contrib/bigquery.py

```diff
@@ -147,6 +158,12 @@ def dataset_exists(self, dataset):
                         fetched_location if fetched_location is not None else 'unspecified',
                         dataset.location))
 
+        except (TimeoutError, BrokenPipeError, IOError) as bq_connection_error:
+            if self.retry_count > self.retry_limit:
+                raise Exception(f"Exceeded max retries for BigQueryClient connection error: {bq_connection_error}") from bq_connection_error
+            self.retry_count += 1
+            self.__initialise_client()
+            self.dataset_exists(dataset)
```

Comment:
How about using retry decorator to simplify the codes?
The implementation below will be useful for you.
https://github.com/spotify/luigi/blob/master/luigi/contrib/gcs.py#L72-L81

Reply:
Thanks, I'll make that change.

Reply:
@jamesmcm Thank you! Looks good to me, but flake8 notifies some formatting problems.

Reply:
Thanks, I fixed that too.